docs/important/relevant_functionalities_de.md

```diff
@@ -0,0 +1,1249 @@
+
+### Teile und herrsche mit Bash
+
+Je länger man eine Anwendung entwickelt, desto mehr Funktionalitäten gelangen in die Anwendung. Genauso kann es sich auch mit der `startup.sh` verhalten. Von einer leichteren Testbarkeit (z. B. mit [BATS](https://github.com/bats-core/bats-core)) abgesehen, sind kleine Ausführungsbestandteile leichter verständlich.
+
+Es ist daher eine gute Idee, Dogu-Funktionalitäten in der `startup.sh` genauso unter Gesichtspunkten der Testbarkeit, Lesbarkeit oder Refaktorisierbarkeit zu gestalten.
+
+#### Shell functions
+
+Hierbei helfen Bash-Funktionen:
+
+```bash
+function setDoguLogLevel() {
+  echo "Mapping dogu specific log level..."
+  currentLogLevel=$(doguctl config --default "WARN" "logging/root")
+  
+  # bilde hier das Loglevel auf die Logkonfiguration des Dogus ab 
+}
+```
+
+Wiederkehrende Funktionen können abstrahiert und parametrisiert werden:
+
+```bash
+function deleteSetupArtefact() {
+  local artefactFile="${1}"
+  rm -f "${artefactFile}"
+}
+
+function cleanUpSetup() {
+  deleteSetupArtefact "/tmp/tracker-extract-3-files.127"
+  deleteSetupArtefact "/usr/share/lib/intermediateFile.bin"
+}
+```
+
+#### Nachvollziehbarkeit während der Startphase
+
+Mit steigender Komplexität ist es eine Idee wert, relevante Schritte mit einem `echo` zu versehen, um im Fehlerfall eine Suche nach dem Fehler in den Logausgaben zu beschleunigen.
+
+
+Zeitgleich gibt es mit dem [Dogu-State-HealthCheck](../core/compendium_de.md#healthchecks) die Möglichkeit, während des Dogu-Starts eigene Ausführungsphasen zu markieren. Zum Beispiel `doguctl state 'aSpecialInstallPhase'` am Anfang oder `doguctl state 'ready'` am Ende des Skripts.
+
+Der Zustand `ready` ist für die Interpretation des Health-Checks zwingend notwendig.
+
+
+#### Auslagerung von Code
+
+Eine weitere Möglichkeit ist die Auslagerung von Funktionen in andere Skriptdateien, die dann mittels `source` in das eigentliche Startskript inkludiert werden. Es sollte allerdings beachtet werden, dass auch ein solches `source`n Fehlschlagen kann und behandelt werden sollte.
+
+```bash
+sourcingExitCode=0
+source /resources/util.sh || sourcingExitCode=$?
+if [[ ${sourcingExitCode} -ne 0 ]]; then
+  echo "ERROR: An error occurred while sourcing /resources/util.sh: Exit code ${sourcingExitCode}"
+  doguctl state "ErrorSourcingUtilSh"
+  sleep 300
+  exit 1
+fi
+```
+
+### Qualitätssicherung von Startskripten
+
+Bei komplexeren Aufgaben kann es sein, dass ein Startskript wie die `startup.sh` an Komplexität gewinnt. Wie jedes andere Programm sollte es getestet und auf Robustheit analysiert werden. Hierbei gibt es mehrere Möglichkeiten. [BATS](https://github.com/bats-core/bats-core) kann im Unit-Test von Shell-Funktionen unterstützen.
+
+Analysewerkzeuge wie [Shellcheck](https://www.shellcheck.net/) können Fehler in der Benutzung von Shell-Aufrufen aufdecken.
```

Comment:
Maybe we should add that passing a shellcheck test is an absolute must-have for dogu scripts, as we will also test this in the Jenkins pipelines.

Reply:
We can only suggest so much. How an external dogu developer develops her CI is completely outside our control

Reply:
I think shellcheck is one of those things we will enforce later, as it is not Dogu-specific and can always be checked on any external dogu code

Reply:
all right, this can be added later